File: greedy/931_minimum_falling_path_sum.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var minFallingPathSum = function(matrix) {

    let ROWS = matrix.length;
    let COLS = matrix[0].length;

    for(let row = 1; row< ROWS; row++){
        for(let col =0; col<COLS; col++){

            let left = Infinity, right = Infinity, center = Infinity;
            center = matrix[row-1][col];

            if(col !== 0){
                left = matrix[row-1][col-1]
            }
            if(col !== COLS-1){
                right = matrix[row-1][col+1]
            }

            matrix[row][col] += Math.min(left, right, center);

        }
        // console.log(matrix)
    }

    return Math.min(... matrix[matrix.length-1])

};

File: greedy/931_minimum_falling_path_sum.rs
impl Solution {
    pub fn min_falling_path_sum(mut matrix: Vec<Vec<i32>>) -> i32 {

        let ROWS = matrix.len();
        let COLS = matrix[0].len();

        for row in 1..ROWS {
            for col in 0..COLS {

                let mut left = i32::MAX;
                let mut right = i32::MAX;
                let mut center = matrix[row-1][col];

                if col != 0 {
                    left = matrix[row-1][col-1];
                }
                if col != COLS-1 {
                    right = matrix[row-1][col+1];
                }

                matrix[row][col] += [left, right, center].iter().cloned().min().unwrap();
            }
        }

        matrix[ROWS - 1].iter().cloned().min().unwrap()

    }
}